packages/flutter/lib/src/widgets/routes.dart

```diff
@@ -1417,6 +1417,57 @@ abstract class ModalRoute<T> extends TransitionRoute<T> with LocalHistoryRoute<T
     return child;
   }
 
+  /// The delegated transition provided to the previous route.
+  DelegatedTransitionBuilder? get delegatedTransition => null;
+
+  /// The delegated transition received from an incoming route.
+  DelegatedTransitionBuilder? receivedTransition;
+
+  /// Transition that matches with the next route. Will be null if there is no
+  /// next route or a next route is not provided. If a non-null value exists,
+  /// then it will be used in place of [buildTransitions].
+  DelegatedTransitionBuilder? get nextRouteTransition {
+    if (receivedTransition != null) {
+      return receivedTransition;
+    } else {
+      return null;
+    }
+  }
+
+  /// Wraps the transitions of the route with a delegated transition received
+  /// from a route being pushed on top of the current route.
+  ///
+  /// This allows the current route to sync its exit transition with the
+  /// entrance transition of the incoming route.
+  ///
+  /// {@tool dartpad}
+  /// This sample shows an app that uses three different page transitions, a
+  /// Material Zoom transition, the standard Cupertino sliding transition, and a
+  /// custom vertical transition. All of the page routes are able to inform the
+  /// previous page how to transition off the screen to sync with the new page.
+  ///
+  /// ** See code in examples/api/lib/widgets/routes/flexible_route_transitions.0.dart **
+  /// {@end-tool}
+  Widget buildFlexTransitions(
+    BuildContext context,
+    Animation<double> animation,
+    Animation<double> secondaryAnimation,
+    Widget child,
+  ) {
+    final Animation<double> proxyAnimation = ProxyAnimation();
+
+    final Animation<double> flexAnimation = (nextRouteTransition == null) ?
+      secondaryAnimation : proxyAnimation;
+
+    final Widget originalTransitions = buildTransitions(context, animation, flexAnimation, child);
+
+    if (nextRouteTransition != null) {
+      return nextRouteTransition!(context, animation, secondaryAnimation, originalTransitions);
+    } else {
+      return originalTransitions;
```

Comment:
If receivedTransition was an empty builder instead of null, could we get rid of all of this mechanism around including it or not including it? Just throwing that out there, not sure if it's a good idea or not yet...